### PR TITLE
Custom Asset Twig/Sanitization

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

- Use sanitized translation functions for text going to `addMessageAfterRedirect` as this method could be used to display HTML content too.
- Prefer `::class` over `getType()`
- PHPDoc fixes
- Migrate connections UI to Twig

### CODE REVIEW

**Review by orthagh — APPROVED** [review]

**Comment by cedric-anne** [comment]
@cconard96 Please rebase to solve conflicts.

**Comment by cconard96** [comment]
> @cconard96 Please rebase to solve conflicts.

done

File: src/Asset/Asset_PeripheralAsset.php [hunk 1]
```diff
@@ -58,15 +59,14 @@ final class Asset_PeripheralAsset extends CommonDBRelation
...
+    public function getForbiddenStandardMassiveAction(): array
     {
-
         $forbidden   = parent::getForbiddenStandardMassiveAction();
         $forbidden[] = 'update';
         return $forbidden;
     }
 
-    public static function getIcon()
+    public static function getIcon(): string
```
Comment (cedric-anne) [inline src/Asset/Asset_PeripheralAsset.php:69]:
```suggestion
    public static function getIcon()
```

File: src/Asset/Asset_PeripheralAsset.php [hunk 2]
```diff
@@ -77,39 +77,38 @@ public static function getIcon()
...
+                'itemtype_peripheral' => $peripheral_item::class,
                 'items_id_peripheral' => $peripheral_item->getID(),
             ]
         );
         return $connections > 0;
     }
 
-
-    public function prepareInputForAdd($input)
+    public function prepareInputForAdd($input): false|array
```
Comment (cedric-anne) [inline src/Asset/Asset_PeripheralAsset.php:96]:
```suggestion
    public function prepareInputForAdd($input)
```

File: src/Asset/Asset_PeripheralAsset.php [hunk 3]
```diff
@@ -192,12 +191,8 @@ public function prepareInputForAdd($input)
         return parent::prepareInputForAdd($input);
     }
 
-
-    public function cleanDBonPurge()
+    public function cleanDBonPurge(): void
```
Comment (cedric-anne) [inline src/Asset/Asset_PeripheralAsset.php:194]:
```suggestion
    public function cleanDBonPurge()
```

File: src/Asset/Asset_PeripheralAsset.php [hunk 4]
```diff
@@ -258,39 +253,33 @@ public function cleanDBonPurge()
...
     }
 
-
     public static function getMassiveActionsForItemtype(
         array &$actions,
         $itemtype,
         $is_deleted = false,
         CommonDBTM $checkitem = null
-    ) {
+    ): void {
```
Comment (cedric-anne) [inline src/Asset/Asset_PeripheralAsset.php:261]:
```suggestion
    ) {
```

File: src/Asset/Asset_PeripheralAsset.php [hunk 5]
```diff
@@ -258,39 +253,33 @@ public function cleanDBonPurge()
...
+        if (in_array($itemtype, $specificities['itemtypes'], true)) {
             $actions[$action_prefix . 'add']    = "<i class='fa-fw fas fa-plug'></i>" . _sx('button', 'Connect');
             $actions[$action_prefix . 'remove'] = _sx('button', 'Disconnect');
         }
         parent::getMassiveActionsForItemtype($actions, $itemtype, $is_deleted, $checkitem);
     }
 
-
-    public static function getRelationMassiveActionsSpecificities()
+    public static function getRelationMassiveActionsSpecificities(): array
```
Comment (cedric-anne) [inline src/Asset/Asset_PeripheralAsset.php:272]:
```suggestion
    public static function getRelationMassiveActionsSpecificities()
```

File: src/Asset/Asset_PeripheralAsset.php [hunk 6]
```diff
@@ -370,116 +338,125 @@ private static function showForAsset(CommonDBTM $asset, $withtemplate = 0)
...
+            $type_class = $itemtype . "Type";
+            $type_table = getTableForItemType($type_class);
+            $type_field = getForeignKeyFieldForTable($type_table);
 
-                echo "<tr class='tab_bg_1'>";
+            if ($_SESSION["glpiis_ids_visible"] || empty($data["name"])) {
+                $linkname = sprintf(__('%1$s (%2$s)'), $linkname, $data["id"]);
+            }
+            $link = $itemtype::getFormURLWithID($data["id"]);
+            $name = "<a href=\"" . htmlspecialchars($link) . "\">" . htmlspecialchars($linkname) . "</a>";
```
Comment (cedric-anne) [inline src/Asset/Asset_PeripheralAsset.php:407]:
```suggestion
            $name = '<a href="' . htmlspecialchars($link) . '">' . htmlspecialchars($linkname) . '</a>';
```

File: src/Asset/Asset_PeripheralAsset.php [hunk 7]
```diff
@@ -525,132 +502,112 @@ private static function showForPeripheral(CommonDBTM $peripheral, $withtemplate
...
+                'serial' => $data["serial"] ?? "-",
+                'otherserial' => $data['otherserial'] ?? '-',
+            ];
+            $linkname = $data["name"];
+            $itemtype = $data['assoc_itemtype'];
+            if ($_SESSION["glpiis_ids_visible"] || empty($data["name"])) {
+                $linkname = sprintf(__('%1$s (%2$s)'), $linkname, $data["id"]);
             }
+            $link = $itemtype::getFormURLWithID($data["id"]);
+            $name = "<a href=\"" . htmlspecialchars($link) . "\">" . htmlspecialchars($linkname) . "</a>";
```
Comment (cedric-anne) [inline src/Asset/Asset_PeripheralAsset.php:563]:
```suggestion
            $name = '<a href="' . htmlspecialchars($link) . '">' . htmlspecialchars($linkname) . '</a>';
```

File: src/Asset/Asset_PeripheralAsset.php [hunk 8]
```diff
@@ -741,8 +697,7 @@ public static function dropdownConnect(
         return $rand;
     }
 
-
-    public function getTabNameForItem(CommonGLPI $item, $withtemplate = 0)
+    public function getTabNameForItem(CommonGLPI $item, $withtemplate = 0): string
```
Comment (cedric-anne) [inline src/Asset/Asset_PeripheralAsset.php:700]:
```suggestion
    public function getTabNameForItem(CommonGLPI $item, $withtemplate = 0)
```

File: src/Asset/Asset_PeripheralAsset.php [hunk 9]
```diff
@@ -768,41 +723,39 @@ public function getTabNameForItem(CommonGLPI $item, $withtemplate = 0)
...
+                    $item::class
                 );
             }
         }
         return '';
     }
 
-
-    public static function displayTabContentForItem(CommonGLPI $item, $tabnum = 1, $withtemplate = 0)
+    public static function displayTabContentForItem(CommonGLPI $item, $tabnum = 1, $withtemplate = 0): bool
```
Comment (cedric-anne) [inline src/Asset/Asset_PeripheralAsset.php:733]:
```suggestion
    public static function displayTabContentForItem(CommonGLPI $item, $tabnum = 1, $withtemplate = 0)
```

File: src/Asset/Asset_PeripheralAsset.php [hunk 10]
```diff
@@ -768,41 +723,39 @@ public function getTabNameForItem(CommonGLPI $item, $withtemplate = 0)
...
             self::showForAsset($item, $withtemplate);
             return true;
         }
 
         return false;
     }
 
-
-    public static function canUnrecursSpecif(CommonDBTM $item, $entities)
+    public static function canUnrecursSpecif(CommonDBTM $item, $entities): bool
```
Comment (cedric-anne) [inline src/Asset/Asset_PeripheralAsset.php:753]:
```suggestion
    public static function canUnrecursSpecif(CommonDBTM $item, $entities)
```

File: src/Asset/Asset_PeripheralAsset.php [hunk 11]
```diff
@@ -868,8 +821,7 @@ public static function canUnrecursSpecif(CommonDBTM $item, $entities)
         return true;
     }
 
-
-    protected static function getListForItemParams(CommonDBTM $item, $noent = false)
+    protected static function getListForItemParams(CommonDBTM $item, $noent = false): array
```
Comment (cedric-anne) [inline src/Asset/Asset_PeripheralAsset.php:824]:
```suggestion
    protected static function getListForItemParams(CommonDBTM $item, $noent = false)
```

File: src/Asset/Asset_PeripheralAsset.php [hunk 12]
```diff
@@ -879,20 +831,19 @@ protected static function getListForItemParams(CommonDBTM $item, $noent = false)
...
+        $table = self::getTable();
         $select = parent::getTypeItemsQueryParams_Select($item);
         $select[] = "$table.is_dynamic AS {$table}_is_dynamic";
 
         return $select;
     }
 
-
-    public static function getTypeName($nb = 0)
+    public static function getTypeName($nb = 0): string
```
Comment (cedric-anne) [inline src/Asset/Asset_PeripheralAsset.php:841]:
```suggestion
    public static function getTypeName($nb = 0)
```

File: src/Asset/Asset_PeripheralAsset.php [hunk 13]
```diff
@@ -879,20 +831,19 @@ protected static function getListForItemParams(CommonDBTM $item, $noent = false)
...
     {
         return _n('Connection', 'Connections', $nb);
     }
 
-    public static function rawSearchOptionsToAdd($itemtype = null)
+    public static function rawSearchOptionsToAdd($itemtype = null): array
```
Comment (cedric-anne) [inline src/Asset/Asset_PeripheralAsset.php:846]:
```suggestion
    public static function rawSearchOptionsToAdd($itemtype = null)
```

File: src/Asset/Asset_PeripheralAsset.php [hunk 14]
```diff
@@ -902,18 +853,15 @@ public static function rawSearchOptionsToAdd($itemtype = null)
...
+                $tab = [...$tab, ...$peripheral::rawSearchOptionsToAdd($itemtype)];
             }
         }
 
         return $tab;
     }
 
     #[Override]
-    public static function getRelationMassiveActionsPeerForSubForm(MassiveAction $ma)
+    public static function getRelationMassiveActionsPeerForSubForm(MassiveAction $ma): int
```
Comment (cedric-anne) [inline src/Asset/Asset_PeripheralAsset.php:864]:
```suggestion
    public static function getRelationMassiveActionsPeerForSubForm(MassiveAction $ma)
```

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**